### ISSUE TITLE
Logitech mx keys S

### ISSUE THREAD

**Issue by forplan0** [JollyFrogs/Backlighter#1]
Hello
I have logitech mx keys S and this solution not working.
Maybe because it's on logitech bolt?
Can you do something about that?

**Comment by JollyFrogs** [comment]
Seems that the MX Keys S has changed backlighting and the way it works, more options. I don't have an MX Keys S unfortunately, so I cannot snoop on the USB interface. However, if you can get the software to turn off and on the backlight, purely via software, and not e.g. by touching it or moving your hand close to it, then that means some command is being sent to do that - in that case, it's possible to make it work, we would just have to sniff the USB 'command' that the software sends to turn on or off the backlight on the MX Keys S. Test it out, let me know if you can get software to turn off and on the backlight, and if so, I can show you how to get the data from the USB interface (using WireShark).

**Comment by forplan0** [comment]
Yes a MX Keys S are little different. They manage to offer setting in options+ to set backlight duration BUT this settings are only for when keyboard is on battery. When you charging is there same 5min backlight duration like in MX keys.

LIke i have settings 20min to backlight off any everything work OK. When my battery is low i need to put cable in and backlight charging is set to 5min!!! 

I don't know how Logitech can be so stupid.

I will check for software on/off but i am not sure if works like that, because when is cable plug-in backlight turn off/on like to switch to hardware settings or something

**Comment by ngfrank45** [comment]
> Seems that the MX Keys S has changed backlighting and the way it works, more options. I don't have an MX Keys S unfortunately, so I cannot snoop on the USB interface. However, if you can get the software to turn off and on the backlight, purely via software, and not e.g. by touching it or moving your hand close to it, then that means some command is being sent to do that - in that case, it's possible to make it work, we would just have to sniff the USB 'command' that the software sends to turn on or off the backlight on the MX Keys S. Test it out, let me know if you can get software to turn off and on the backlight, and if so, I can show you how to get the data from the USB interface (using WireShark).

Hello, 
I am actually trying to do this but for Logitech MX Mechanical with Bolt Receiver, so I modified this part of code and changed productId to 0xc548 (Logi bolt receiver instead of Logi unifying receiver) : 

`FilterDeviceDefinition(vendorId: 0x046d, productId: 0xc548, label: "Logitech - USB Receiver", usagePage: 65280)`

![Screenshot 2025-01-12 092119](https://github.com/user-attachments/assets/f9c273a9-b095-4d45-8c4e-cd544e15cbc9)

Will it still work no matter what label I give as an argument ? 
Because I have no idea if this one is the right one or is there is a specific label to give. 
I also had other HID named "Logitech - USB Receiver", but they had a different usagePage and I preferred to pick the same as you(0xFF00 / 65280). 
I also didn't change deviceDefinitions.First() and picked the first device from the list(there was a count of 2 devices)

But I also have another problem, I sniffed packages from usb on Wireshark but I can't find the good bytes to send no matter how hard I try.
So I clicked the disconnect button on logi options software then stopped Wireshark, then I had some around some potentiel candidates packages who stand out, do you remember if you used a package with URB_BULK, URB_INTERRUPT, SET_REPORT or Rcvd handle notification ? It could even be none of them as we are not using the same exact keyboard.

![Screenshot 2025-01-12 114314](https://github.com/user-attachments/assets/3cb5c54a-fcbf-477d-889b-ba0d88120f91)

Personally I don't think it's Rcvd handle notification or URB_BULK because there is hundreds of them and it's mostly junk, but it could probably be 56 SET_REPORT Request or one of the two 27 URB_INTERRUPT in in as they are rarer and appears mostly when I shut down the backlight on the software, also they are sent from the host to a destination and not the other way around.
But still I don't know which bytes to use, and I can't figure it out. Could you help me please ? 

Here is the data for thoses requests : 

27 URB_INTERRUPT in (first one) :

![Screenshot 2025-01-12 115157](https://github.com/user-attachments/assets/bba089e9-534b-449f-bb2a-b8d1552dd0b5)


56 SET_REPORT Request (Frame) : 

![Screenshot 2025-01-12 115629](https://github.com/user-attachments/assets/3b68ed7b-42e0-474f-a31c-10e6acab8b3d)

56 SET_REPORT Request (USB Control) : 

![Screenshot 2025-01-12 115726](https://github.com/user-attachments/assets/72d22d36-c98d-4ee3-941b-7d4f91d69824)


27 URB_INTERRUPT in (second one) :

![Screenshot 2025-01-12 115814](https://github.com/user-attachments/assets/5338b1fe-d20c-4c8c-8bed-d50920c4004f)

(ps: right now I am only looking for the signal off bytes, because I might be able to reproduce the steps and find the signal on bytes too if I can at least find the signal off bytes)

**Comment by JollyFrogs** [comment]
Hi Mate,

The way I figured it out is via WireShark -> USB capture. Then just check
what it sends when the light is turned on or off. That gave me the packets
I needed.

Cheers,
Alex

**Comment by JollyFrogs** [comment]
awesome, sounds like a working solution to me - whatever works usually does
it :) I like how hackish it is, it doesn't hide much :D Well done mate!
